File: backend/app/service_area.py
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import TypeAlias

from fastapi import HTTPException, status


SERVICE_AREA_ERROR = "Reports can only be submitted within Kuala Lumpur."
Coordinate: TypeAlias = tuple[float, float]
Ring: TypeAlias = list[Coordinate]
Polygon: TypeAlias = list[Ring]


def _boundary_path() -> Path:
    return Path(__file__).resolve().parents[2] / "src" / "data" / "kuala-lumpur-boundary.geojson"


@lru_cache(maxsize=1)
def _load_polygons() -> list[Polygon]:
    payload = json.loads(_boundary_path().read_text(encoding="utf-8"))
    features = payload.get("features", [])
    polygons: list[Polygon] = []

    for feature in features:
        geometry = feature.get("geometry", {})
        geometry_type = geometry.get("type")
        coordinates = geometry.get("coordinates", [])

        if geometry_type == "Polygon":
            polygons.append(_to_polygon(coordinates))
        elif geometry_type == "MultiPolygon":
            polygons.extend(_to_polygon(polygon) for polygon in coordinates)

    if not polygons:
        raise RuntimeError("Kuala Lumpur service-area boundary could not be loaded.")

    return polygons


def _to_polygon(coordinates: list) -> Polygon:
    return [
        [(float(longitude), float(latitude)) for longitude, latitude in ring]
        for ring in coordinates
    ]


def _point_on_segment(
    longitude: float,
    latitude: float,
    start: Coordinate,
    end: Coordinate,
) -> bool:
    start_longitude, start_latitude = start
    end_longitude, end_latitude = end
    cross_product = (latitude - start_latitude) * (end_longitude - start_longitude) - (
        longitude - start_longitude
    ) * (end_latitude - start_latitude)

    if abs(cross_product) > 1e-10:
        return False

    return (
        min(start_longitude, end_longitude) - 1e-10
        <= longitude
        <= max(start_longitude, end_longitude) + 1e-10
        and min(start_latitude, end_latitude) - 1e-10
        <= latitude
        <= max(start_latitude, end_latitude) + 1e-10
    )


def _point_in_ring(longitude: float, latitude: float, ring: Ring) -> bool:
    inside = False
    previous = ring[-1]

    for current in ring:
        if _point_on_segment(longitude, latitude, previous, current):
            return True

        current_longitude, current_latitude = current
        previous_longitude, previous_latitude = previous
        crosses_latitude = (current_latitude > latitude) != (previous_latitude > latitude)

        if crosses_latitude:
            intersection_longitude = (
                (previous_longitude - current_longitude)
                * (latitude - current_latitude)
                / (previous_latitude - current_latitude)
                + current_longitude
            )
            if longitude < intersection_longitude:
                inside = not inside

        previous = current

    return inside


def _point_in_polygon(longitude: float, latitude: float, polygon: Polygon) -> bool:
    if not polygon or not _point_in_ring(longitude, latitude, polygon[0]):
        return False

    return not any(_point_in_ring(longitude, latitude, hole) for hole in polygon[1:])


def is_within_service_area(latitude: float, longitude: float) -> bool:
    return any(
        _point_in_polygon(longitude, latitude, polygon)
        for polygon in _load_polygons()
    )


def ensure_within_service_area(latitude: float, longitude: float) -> None:
    if not is_within_service_area(latitude, longitude):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=SERVICE_AREA_ERROR,
        )

File: src/lib/serviceArea.ts
import boundaryGeoJson from '@/data/kuala-lumpur-boundary.geojson?raw'
import type { LocationPoint } from '@/types/report'

type Coordinate = [number, number]
type Ring = Coordinate[]
type Polygon = Ring[]
type Geometry =
  | { type: 'Polygon'; coordinates: Polygon }
  | { type: 'MultiPolygon'; coordinates: Polygon[] }

interface BoundaryFeature {
  type: 'Feature'
  geometry: Geometry
}

interface BoundaryFeatureCollection {
  type: 'FeatureCollection'
  features: BoundaryFeature[]
}

export const SERVICE_AREA_ERROR = 'Reports can only be submitted within Kuala Lumpur.'
export const SERVICE_AREA_LABEL = 'Kuala Lumpur service area'

const boundary = JSON.parse(boundaryGeoJson) as BoundaryFeatureCollection
const serviceAreaPolygons = boundary.features.flatMap((feature) =>
  feature.geometry.type === 'Polygon'
    ? [feature.geometry.coordinates]
    : feature.geometry.coordinates,
)
const allBoundaryCoordinates = serviceAreaPolygons.flat(2)

function isPointOnSegment(
  longitude: number,
  latitude: number,
  [startLongitude, startLatitude]: Coordinate,
  [endLongitude, endLatitude]: Coordinate,
) {
  const crossProduct =
    (latitude - startLatitude) * (endLongitude - startLongitude) -
    (longitude - startLongitude) * (endLatitude - startLatitude)

  if (Math.abs(crossProduct) > 1e-10) {
    return false
  }

  return (
    Math.min(startLongitude, endLongitude) - 1e-10 <= longitude &&
    longitude <= Math.max(startLongitude, endLongitude) + 1e-10 &&
    Math.min(startLatitude, endLatitude) - 1e-10 <= latitude &&
    latitude <= Math.max(startLatitude, endLatitude) + 1e-10
  )
}

function isPointInRing(longitude: number, latitude: number, ring: Ring) {
  let inside = false
  let previous = ring.at(-1)

  if (!previous) {
    return false
  }

  for (const current of ring) {
    if (isPointOnSegment(longitude, latitude, previous, current)) {
      return true
    }

    const [currentLongitude, currentLatitude] = current
    const [previousLongitude, previousLatitude] = previous
    const crossesLatitude = (currentLatitude > latitude) !== (previousLatitude > latitude)

    if (crossesLatitude) {
      const intersectionLongitude =
        ((previousLongitude - currentLongitude) * (latitude - currentLatitude)) /
          (previousLatitude - currentLatitude) +
        currentLongitude

      if (longitude < intersectionLongitude) {
        inside = !inside
      }
    }

    previous = current
  }

  return inside
}

function isPointInPolygon(longitude: number, latitude: number, polygon: Polygon) {
  const [outerRing, ...holes] = polygon

  if (!outerRing || !isPointInRing(longitude, latitude, outerRing)) {
    return false
  }

  return !holes.some((hole) => isPointInRing(longitude, latitude, hole))
}

export function isWithinServiceArea(point: Pick<LocationPoint, 'latitude' | 'longitude'>) {
  return serviceAreaPolygons.some((polygon) =>
    isPointInPolygon(point.longitude, point.latitude, polygon),
  )
}

export const serviceAreaBoundaryPolygons = serviceAreaPolygons.map((polygon) => {
  const firstCoordinate = polygon[0]?.[0]
  const id = firstCoordinate
    ? `service-area-${firstCoordinate[0]}-${firstCoordinate[1]}`
    : SERVICE_AREA_LABEL

  return {
    id,
    positions: polygon.map((ring) =>
      ring.map(([longitude, latitude]) => [latitude, longitude] as [number, number]),
    ),
  }
})

const worldMaskRing: [number, number][] = [
  [90, -180],
  [90, 180],
  [-90, 180],
  [-90, -180],
]

export const serviceAreaMaskPositions = [
  worldMaskRing,
  ...serviceAreaPolygons
    .map((polygon) => polygon[0])
    .filter((ring): ring is Ring => Boolean(ring))
    .map((ring) =>
      ring.map(([longitude, latitude]) => [latitude, longitude] as [number, number]),
    ),
]

const longitudes = allBoundaryCoordinates.map(([longitude]) => longitude)
const latitudes = allBoundaryCoordinates.map(([, latitude]) => latitude)
const boundsPadding = 0.025

export const serviceAreaMapBounds: [[number, number], [number, number]] = [
  [Math.min(...latitudes) - boundsPadding, Math.min(...longitudes) - boundsPadding],
  [Math.max(...latitudes) + boundsPadding, Math.max(...longitudes) + boundsPadding],
]
